lib/config/applyrule.hpp

```diff
@@ -15,11 +16,53 @@ namespace icinga
 /**
  * @ingroup config
  */
-class ApplyRule
+class ApplyRule : public SharedObject
 {
 public:
+	DECLARE_PTR_TYPEDEFS(ApplyRule);
+
+	struct PerHost
+	{
+		std::set<ApplyRule::Ptr> ForHost;
+		std::unordered_map<String /* service */, std::set<ApplyRule::Ptr>> ForServices;
+	};
+
+	struct PerSourceType
+	{
+		std::unordered_map<Type* /* target type */, std::vector<ApplyRule::Ptr>> Regular;
+		std::unordered_map<String /* host */, PerHost> Targeted;
+	};
+
+	struct PtrCmp
+	{
+		bool operator()(const String* lhs, const String* rhs) const
+		{
+			return *lhs < *rhs;
+		}
+
+		bool operator()(const std::pair<const String *, const String *>& lhs, const std::pair<const String *, const String *>& rhs) const
+		{
+			return *lhs.first < *rhs.first || *lhs.second < *rhs.second;
+		}
```

Comment:
By the way, this one does not work properly. It says that `{"a", "b"} < {"b", "a"}` **and** `{"b", "a"} < {"a", "b"}`.